test/test_deployer.go

```diff
@@ -0,0 +1,117 @@
+package main
+
+import (
+	"fmt"
+	"time"
+
+	c "github.com/flynn/flynn/Godeps/_workspace/src/github.com/flynn/go-check"
+	ct "github.com/flynn/flynn/controller/types"
+)
+
+type DeployerSuite struct {
+	Helper
+}
+
+var _ = c.Suite(&DeployerSuite{})
+
+func (s *DeployerSuite) createDeployment(t *c.C, strategy string) *ct.Deployment {
+	app, release := s.createApp(t)
+	app.Strategy = strategy
+	s.controllerClient(t).UpdateApp(app)
+
+	jobStream := make(chan *ct.JobEvent)
+	scale, err := s.controllerClient(t).StreamJobEvents(app.Name, 0, jobStream)
+	t.Assert(err, c.IsNil)
+	defer scale.Close()
+
+	t.Assert(s.controllerClient(t).PutFormation(&ct.Formation{
+		AppID:     app.ID,
+		ReleaseID: release.ID,
+		Processes: map[string]int{"printer": 2},
+	}), c.IsNil)
+
+	waitForJobEvents(t, scale, jobStream, jobEvents{"printer": {"up": 2}})
+
+	// create a new release for the deployment
+	release.ID = ""
+	t.Assert(s.controllerClient(t).CreateRelease(release), c.IsNil)
+
+	deployment, err := s.controllerClient(t).CreateDeployment(app.ID, release.ID)
+	t.Assert(err, c.IsNil)
+	return deployment
+}
+
+func waitForDeploymentEvents(t *c.C, stream chan *ct.DeploymentEvent, expected []*ct.DeploymentEvent) {
+	// wait for an event with no release to mark the end of the deployment,
+	// collecting events along the way
+	events := []*ct.DeploymentEvent{}
+loop:
+	for {
+		select {
+		case e := <-stream:
+			events = append(events, e)
+			if e.Status == "complete" {
+				break loop
+			}
+		case <-time.After(5 * time.Second):
+			t.Fatal("timed out waiting for deployment event")
+		}
+	}
+	compare := func(t *c.C, i *ct.DeploymentEvent, j *ct.DeploymentEvent) {
+		fmt.Println("Comparing", i, j)
```

Comment:
This should call `debug` so that it's only outputted on failure and includes timestamps